src/d3a/d3a_core/simulation.py

```diff
@@ -363,13 +371,19 @@ def _execute_simulation(self, slot_resume, tick_resume, console=None):
 
                 self.area.tick_and_dispatch()
                 self.area.update_area_current_tick()
+                if self.simulation_config.external_connection_enabled and \
+                        self.external_tick_counter.is_it_time_for_external_tick(tick_no):
```

Comment:
So the external_tick_counter in the simulation.py is only useful for calculating if the global_objects need to be updated? There is an alternative solution to that, which is to make the external_tick_counter a part of the `GlobalStatistics` class , and rename the class to `ExternalConnectionGlobalStatistics`. That way you will have only one counter for the whole simulation, and not separate counters per strategy. 
Also, this has the advantage that the `if self.simulation_config.external_connection_enabled and self.external_tick_counter.is_it_time_for_external_tick(tick_no)` checks can be moved inside the update() method, therefore all you will have to do from the caller code is `update()`, and the simulation.py code will be much cleaner.

Reply:
@spyrostz unfortunately not, the `external_tick_counter` is also used in the `ExternalMixin._dispatch_event_tick_to_external_agent` were it was originating from. Now i also needed for deciding whether the global updates need to be updated.

Reply:
Yes, but this is easy to circumvent if you implement the singleton `ExternalConnectionGlobalStatistics`. You can expose a public method `is_it_time_for_external_tick` on the singleton (or, even better, a property. Since you update the global statistics object on every tick, it is rather easy to buffer the 'is this tick supposed to update external agents' information) and use the buffered version, instead of directly calling the `is_it_time_for_external_tick`. 
In general, I think this singleton would be a very good first step to a change that I have been thinking of for quite some time. At the moment, we have scattered across our code various objects that contain state for the whole simulation. The ideal way to deal with this is to split this state into different objects (e.g. ConfigSettings for configuration, ExternalConnectionState for external connection-specific variables) and then create a wrapper class that will instantiate all these objects, with this class only being a singleton (or, even better, constructed in the Simulation class, which we know it is only one per simulation). Before we do something like this though, we have to isolate the state, that's why I am insisting to move the external_tick_counter inside the same class that handles other parts of external connection state. Even if you think about it conceptually, even if we instantiate the ExternalTickCounter multiple times, the ticks are essentially centralized (tick has the same value for all areas, before or after the event_tick)

Reply:
Ok, understood and agreed. I will add another singleton then (on top of global_statistics that i am implementing right now)

Reply:
Awesome, just one last comment, the ExternalTickCounter does not need to be a singleton, since it is only going to be constructed by the ExternalConnectionGlobalStatistics singleton it can be a normal class.

Reply:
Thanks for the clarification! Would have done it wrong.